### PR TITLE
[AMDGPU] Combine or remove redundant waitcnts at the end of each MBB

### PR DESCRIPTION
**Pull request by jayfoad**
Call generateWaitcnt unconditionally at the end of
SIInsertWaitcnts::insertWaitcntInBlock. Even if we don't need to
generate a new waitcnt instruction it has the effect of combining or
removing redundant waitcnts that were already present. Tests show
various small improvements in waitcnt placement.

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.is.private.ll [hunk 1]
```diff
@@ -131,8 +131,6 @@ define amdgpu_kernel void @is_private_sgpr(ptr %ptr) {
 ; GFX11-NEXT:    global_store_b32 v[0:1], v0, off dlc
 ; GFX11-NEXT:    s_waitcnt_vscnt null, 0x0
 ; GFX11-NEXT:  .LBB1_2: ; %bb1
-; GFX11-NEXT:    s_nop 0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.is.private.ll:134]: Dont' understand this change, no waits changed?

Comment (jayfoad) [reply llvm/test/CodeGen/AMDGPU/GlobalISel/llvm.amdgcn.is.private.ll:134]: SIInsertWaitcnts only inserts that s_nop/s_sendmsg sequence if it thinks that vscnt might be non-zero at the s_endpgm. I think what happened here is that the pass now takes notice of the "s_waitcnt_vscnt 0" at the end of the previous MBB. You're right that it didn't get changed, but the pass's bookkeeping got updated with the knowledge that vscnt must be zero after that.

**Review by arsenm — APPROVED** [review]

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 2]
```diff
@@ -2355,9 +2327,22 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
...
+      isPreheaderToFlush(Block, ScoreBrackets)) {
+    if (ScoreBrackets.hasPendingEvent(LOAD_CNT))
+      Wait.LoadCnt = 0;
+    if (ScoreBrackets.hasPendingEvent(SAMPLE_CNT))
+      Wait.SampleCnt = 0;
+    if (ScoreBrackets.hasPendingEvent(BVH_CNT))
+      Wait.BvhCnt = 0;
+  }
+
+  // Combine or remove any redundant waitcnts ar the end of the block.
```
Comment (jwanggit86) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:2343]: "ar the end" => "at the end"

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp [hunk 3]
```diff
@@ -2355,9 +2327,22 @@ bool SIInsertWaitcnts::insertWaitcntInBlock(MachineFunction &MF,
     ++Iter;
   }
 
+  // Flush the LOADcnt, SAMPLEcnt and BVHcnt counters at the end of the block if
+  // needed.
+  AMDGPU::Waitcnt Wait;
```
Comment (jwanggit86) [inline llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:2332]: Move the variable "Wait" inside the if?

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp (hunk 2 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp]: Done

File: llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp (hunk 3 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp:2332]: No. It's used outside the if.

**Pull request merged**